lib/ret_web/channels/hub_channel.ex

```diff
@@ -455,6 +455,54 @@ defmodule RetWeb.HubChannel do
     {:noreply, socket}
   end
 
+  def handle_in("list_entities", _, socket) do
+    hub = socket |> hub_for_socket
+    entities = Ret.list_entities(hub.hub_id)
+    {:reply, {:ok, EntityView.render("index.json", %{entities: entities})}, socket}
+  end
+
+  def handle_in("save_entity_state", params, socket) do
+    params = parse(params)
+
+    with {:ok, hub} <- authorize(socket, :write_entity_state),
+         {:ok, %{entity: entity}} <- Ret.create_entity(hub, params) do
+      entity = Repo.preload(entity, [:sub_entities])
+      broadcast!(socket, "entity_state_saved", EntityView.render("show.json", %{entity: entity}))
+      {:reply, :ok, socket}
+    else
+      {:error, reason} ->
+        reply_error(socket, reason)
+    end
+  end
+
+  def handle_in("update_entity_state", %{"update_message" => update_message} = params, socket) do
+    params = parse(params)
+
+    with {:ok, hub} <- authorize(socket, :write_entity_state),
+         {:ok, _} <- Ret.insert_or_update_sub_entity(hub, params) do
```

Comment:
Confirming. Are sub-entity insertions and update both possible for `"update_entity_state"`?

Reply:
Yep! If one exists, it should be updated. If it's a new one, it should be inserted.